### PR TITLE
ShizaProject graphql indexer

### PR DESCRIPTION
**Pull request by dm3ch**
Here's attempt of rewriting ShizaProject indexer to use graphql api of site.

/closes https://github.com/Jackett/Jackett/issues/11712

### CODE REVIEW

File: src/Jackett.Common/Indexers/ShizaProject.cs [hunk 1]
```diff
@@ -45,170 +48,141 @@ internal class ShizaProject : BaseWebIndexer
...
+        
+        private async Task<IEnumerable<ReleaseInfo>> FetchNewReleases()
         {
-            var releases = new List<ReleaseInfo>();
-            var uri = new Uri(url);
-            var result = await RequestWithCookiesAndRetryAsync(url);
-
-            try
+            logger.Info(GraphqlEndpointUrl);
+            using var graphQLClient = new GraphQLHttpClient(GraphqlEndpointUrl, new NewtonsoftJsonSerializer());
```
Comment (dm3ch) [inline src/Jackett.Common/Indexers/ShizaProject.cs:113]: To continue development I have to make GraphQL client works with proxy.

Constructor of GraphQLHttpClient could use HttpClient option to use configured HttpClient (https://github.com/graphql-dotnet/graphql-client/blob/master/src/GraphQL.Client/GraphQLHttpClient.cs#L66).
As far as I understands in indexer we should already have configured httpClient, so my question is how to get it to initialize GraphQLHttpClient?

@ngosang Maybe you can consult me as one of authors of https://github.com/Jackett/Jackett/blob/master/src/Jackett.Common/Utils/Clients/HttpWebClient2.cs

**Review by dm3ch — COMMENTED** [review]
@ngosang Could you please consult me about some complex C# things? :)

**Comment by dm3ch** [comment]
Status: Currently this PR works completely fine for me.

Missing:
- Pretty refactor
- Support of proxy for graphql requests

**Comment by dm3ch** [comment]
@garfield69 Now this indexer is fully functional.

The only problem I have now is that I need somehow get httpClient instance to use it in GraphQL initialisation to make proxy settings working. Maybe you can help me and assign some C# reviewer that would be able to consult me?

**Comment by garfield69** [comment]
> @garfield69 Now this indexer is fully functional.
> 
> The only problem I have now is that I need somehow get httpClient instance to use it in GraphQL initialisation to make proxy settings working. Maybe you can help me and assign some C# reviewer that would be able to consult me?

I'm not a C# dev myself and have little knowledge about proxies, httpclients and the internals of Jackett. While there are some casual C# contributors I don't know their capabilities and interests in this project.
@ngosang  has been the most reliable go-to-guy for this kind of stuff, but this year he has pulled back a bit with other commitments. I suggest you email direct to catch his attention as posts in github are not enough atm.  Try ngosang@hotmail.es

**Comment by ngosang** [comment]
@dm3ch Thank you! I like the new code and I don't have problems adding the new dependency.

Yes, we are using an HttpClient but we have 2 different implementations. One for .net framework/ mono and other for .Net Core. It's possible to do a refactor in the classes `HttpWebClient, HttpWebClient2 and WebClient` to expose the httpclient but it can be tricky.

IMHO I think the best way will be to use the GrapthQL library only for mapping request/response and not using network capabilities. You can use the the existing methods in `BaseIndexer` like `RequestWithCookiesAsync`. These methods don't do parsing. You can send a String body and receive a String response. I don't know if the GraphQL library supports to parse a String GraphQL response into C# objects.

**Comment by dm3ch** [comment]
@garfield69 @ngosang I tried to find way to use graphql-client library only for response parsing, but it seems to complex for me (maybe because I' a noob in C#).

So for now I have implemented passing of httpClient from `webClient`, cause it seems much easier for me (maybe because my code have not so good quality :) ). So now ShizaProject grapghql indexer uses proxy on both httpWebClients implementations (I tested that).

I think this PR is ready for review.

P.S. @ngosang Why mono 5.0 is incompatible with httpWebClient? I didn't found any information about that in code. :(
P.S.S. Maybe it would be better to separate changes of WebClient, httpWebClient and httpWebClient2 to separate PR?

**Comment by ngosang** [comment]
> I tried to find way to use graphql-client library only for response parsing, but it seems to complex for me (maybe because I' a noob in C#).

According to [this](https://dgraph.io/docs/graphql/api/requests/), the request is in GraphQL format but the response is plain JSON. You can remove the GrapthQL dependency completely and parse the JSON response. We are already doing that in most indexers.
Take a look at this example. Simple HTTP request and convert String response into Object with JSONConvert => https://github.com/Jackett/Jackett/blob/6a213c6eabd0264392c48564ff16a9035fd13ca4/src/Jackett.Common/Indexers/ThePirateBay.cs#L209
You can even generate the response Objects with https://app.quicktype.io/?l=csharp

Keep this PR open as reference if you want and open another without GrapQL dependency.

> Why mono 5.0 is incompatible with httpWebClient? I didn't found any information about that in code. :(

I'm not 100% sure but I think we are compatible old versions of Mono (3 and 4) because we support really weird OS like FreeBSD, NAS, etc... They don't have native .Net Core neither Mono 5 neither Docker support...

I don't have much time lately to review and test you code carefully so I'm afraid we can break something important. In the past we broke the updater with a similar change. That's why I'm proposing to not use GraphQL library, to avoid touching httpclient code.

**Comment by dm3ch** [comment]
@ngosang Rewrote indexer one more time as you suggested.
So it's ready for review again

P.S. The most tricky part for me was request forming not response parsing. :)

**Review by ngosang — APPROVED** [review]
Nice. Sorry for the wasted time.

**Comment by dm3ch** [comment]
@ngosang Thank you for help with this PR

**Pull request merged**